titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java

```diff
@@ -0,0 +1,129 @@
+                .endpointResolver(endpointResolver)
+                // We rely on the RxRestClient for timeout handling
+                .timeUnit(TimeUnit.MILLISECONDS)
+                .requestTimeout(titusRegistryClientConfiguration.getRegistryTimeoutMs())
+                .retryCount(titusRegistryClientConfiguration.getRegistryRetryCount())
+                .retryDelay(titusRegistryClientConfiguration.getRegistryRetryDelayMs())
+                .noRetryStatuses(new HashSet<>(Collections.singletonList(HttpResponseStatus.NOT_FOUND)));
+        if (titusRegistryClientConfiguration.isSecure()) {
+            try {
+                builder.sslEngineFactory(new NettySslContextEngineFactory(SslContextBuilder.forClient().build()));
+            } catch (SSLException e) {
+                logger.error("Unable configure Docker registry client SSL context: {}", e.getMessage());
+                throw new RuntimeException("Error configuring SSL context", e);
+            }
+        }
+        this.restClient = builder.build();
+    }
+
+    /**
+     * Gets the Docker Version 2 Schema 2 Content Digest for the provide repository and reference. The
+     * reference may be an image tag or digest value. If the image does not exist or another error is
+     * encountered, an onError value is emitted.
+     */
+    public Single<String> getImageDigest(String repository, String reference) {
+        return registryRequestWithErrorHandling(restClient.doGET(buildRegistryUri(repository, reference), headers, TypeProviders.ofEmptyResponse()))
+                .map(RxHttpResponse::getHeaders)
+                .flatMap(stringListMap -> {
+                    if (stringListMap.containsKey(dockerDigestHeaderKey)) {
+                        return Observable.from(stringListMap.get("Docker-Content-Digest"));
+                    }
+                    return Observable.error(new TitusRegistryException(TitusRegistryException.ErrorCode.MISSING_HEADER, "Missing required header " + dockerDigestHeaderKey));
+                }).first().toSingle();
+    }
+
+    private String buildRegistryUri(String repository, String reference) {
+        return "/v2/" + repository + "/manifests/" + reference;
+    }
+
+    /**
+     * Wraps an observable registry request with timeouts and error handling.
+     */
+    private <T> Observable<T> registryRequestWithErrorHandling(Observable<T> obs) {
+        return obs.timeout(titusRegistryClientConfiguration.getRegistryTimeoutMs(), TimeUnit.MILLISECONDS)
+                .onErrorResumeNext(throwable -> {
+                    if (throwable instanceof RxRestClientException) {
+                        if (((RxRestClientException)throwable).getStatusCode() == HttpResponseStatus.NOT_FOUND.code()) {
+                            return Observable.error(
+                                    new TitusRegistryException(TitusRegistryException.ErrorCode.IMAGE_NOT_FOUND,
+                                    "Image does not exist in registry"));
```

Comment:
pass in image name so this message will be more useful

Reply:
Will do.